internal/keepers/service.go

```diff
@@ -132,15 +143,90 @@ func (s *onDemandUpkeepService) CheckUpkeep(ctx context.Context, keys ...types.U
 func (s *onDemandUpkeepService) start() {
 	// TODO: if this process panics, restart it
 	go s.cacheCleaner.Run(s.cache)
+	go func() {
+		if err := s.runSamplingUpkeeps(); err != nil {
+			s.logger.Fatal(err)
+		}
+	}()
 }
 
 func (s *onDemandUpkeepService) stop() {
 	close(s.stopProcs)
 	close(s.cacheCleaner.stop)
 }
 
-func (s *onDemandUpkeepService) parallelCheck(ctx context.Context, keys []types.UpkeepKey, filters ...func(types.UpkeepKey) bool) ([]*types.UpkeepResult, error) {
-	samples := newSyncedArray[*types.UpkeepResult]()
+func (s *onDemandUpkeepService) runSamplingUpkeeps() error {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	heads := make(chan types.BlockKey, 1)
+	defer close(heads)
+
+	// Start the sampling upkeep process for heads
+	go func() {
+		for head := range heads {
+			s.handleIncomingHead(ctx, head)
+		}
+	}()
+
+	// Cancel context when receiving the stop signal
+	go func() {
+		<-s.stopProcs
+		cancel()
+	}()
+
+	return s.headSubscriber.OnNewHead(ctx, func(blockKey types.BlockKey) {
+		// This is needed in order to do not block the process when a new head comes in.
+		// The running upkeep sampling process should be finished first before starting
+		// sampling for the next head.
```

Comment:
question about channel: when we are doing `s.handleIncomingHead(ctx, head)` for a particular head, is it removed from the channel or will it be removed once that call finishes? 

What i want to confirm is that if this behaviour is possible

- Head no. 10 is put into channel
- handleIncomingHead(10) is called by the other go routine
- Head no 11. is put into the channel (Not sure about this)
- Head no 12. comes and is skipped since channel is full
- Head no 13. comes and is skipped since channel is full
- handleIncomingHead(10) completes
- handleIncomingHead(11) is called

Reply:
Exactly how it works right now since the head channel has buffer == 1. If I make this channel sync (buffer == 0), then this scenario would be changed to:

- Head no. 10 is put into channel
- handleIncomingHead(10) is called by the other go routine
- Head no 11. comes and is skipped since channel is full (reader is not able to read yet)
- Head no 12. comes and is skipped since channel is full (reader is not able to read yet)
- handleIncomingHead(10) completes
- Head no 13. comes and handleIncomingHead(13) is called

Reply:
ok i think we need to design this better. 

In first case we'll end up checking on old block (block 11) once `handleIncomingHead` finishes
In second case we will have to wait for a new block once `handleIncomingHead` finishes

Both aren't ideal

Reply:
@begmaroman maybe a new struct would be useful that behaves a lot like Go's implementation of `Ticker`. That is, if a new head has come in and the receiver is slow to consume, older heads are dropped and the latest is always waiting. If the consumer is fast, the cadence is just that of incoming heads.